File: app/src/main/java/com/musicplayer/utils/CacheMetadataManager.java
package com.musicplayer.utils;

import android.content.Context;
import android.content.SharedPreferences;
import com.musicplayer.model.Song;
import org.json.JSONObject;

public class CacheMetadataManager {
    private static final String PREF_NAME = "cache_metadata";

    public static void saveMetadata(Context context, Song song) {
        if (context == null || song == null || song.getUrl() == null) return;
        
        SharedPreferences prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        try {
            JSONObject obj = new JSONObject();
            obj.put("id", song.getId());
            obj.put("title", song.getTitle());
            obj.put("artist", song.getArtist());
            obj.put("url", song.getUrl());
            obj.put("duration", song.getDuration());
            obj.put("thumbnailUrl", song.getThumbnailUrl());
            
            // Usamos la URL original como clave principal, ya que es lo que se envía al CacheDataSource
            prefs.edit().putString(song.getUrl(), obj.toString()).apply();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static Song getMetadata(Context context, String url) {
        if (context == null || url == null) return null;
        
        SharedPreferences prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        String json = prefs.getString(url, null);
        if (json == null) return null;
        
        try {
            JSONObject obj = new JSONObject(json);
            Song song = new Song(
                obj.optString("id"),
                obj.optString("title"),
                obj.optString("artist"),
                obj.optString("url"),
                obj.optInt("duration"),
                obj.optString("thumbnailUrl")
            );
            song.setLocal(false); // Es remotamente local, pero es cacheado
            return song;
        } catch (Exception e) {
            return null;
        }
    }
}

File: app/src/main/java/com/musicplayer/utils/CacheMetadataManager.kt
package com.musicplayer.utils

import android.content.Context
import com.musicplayer.model.Song
import org.json.JSONObject

object CacheMetadataManager {
    private const val PREF_NAME = "cache_metadata"

    @JvmStatic
    fun saveMetadata(context: Context?, song: Song?) {
        if (context == null || song == null || song.url == null) return
        
        val prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
        try {
            val obj = JSONObject()
            obj.put("id", song.id)
            obj.put("title", song.title)
            obj.put("artist", song.artist)
            obj.put("url", song.url)
            obj.put("duration", song.duration)
            obj.put("thumbnailUrl", song.thumbnailUrl)
            
            prefs.edit().putString(song.url, obj.toString()).apply()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    @JvmStatic
    fun getMetadata(context: Context?, url: String?): Song? {
        if (context == null || url == null) return null
        
        val prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
        val json = prefs.getString(url, null) ?: return null
        
        return try {
            val obj = JSONObject(json)
            Song(
                obj.optString("id"),
                obj.optString("title"),
                obj.optString("artist"),
                obj.optString("url"),
                obj.optLong("duration", 0),
                obj.optString("thumbnailUrl")
            ).apply {
                isLocal = false
            }
        } catch (e: Exception) {
            null
        }
    }

    @JvmStatic
    fun getAllCachedSongs(context: Context?): List<Song> {
        val songs = mutableListOf<Song>()
        if (context == null) return songs
        val prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
        for (entry in prefs.all.entries) {
            try {
                val json = entry.value as String
                val obj = JSONObject(json)
                songs.add(
                    Song(
                        obj.optString("id"),
                        obj.optString("title"),
                        obj.optString("artist"),
                        obj.optString("url"),
                        obj.optLong("duration", 0),
                        obj.optString("thumbnailUrl")
                    ).apply {
                        isLocal = false
                    }
                )
            } catch (ignored: Exception) {}
        }
        return songs
    }
}
